<a href="https://colab.research.google.com/github/oguzaltuntas/permutasyon/blob/master/Perm%C3%BCtasyon.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [41]:
 def factorialCalc(i): # permütasyon hesaplamanın temeli olan faktöriyel hesaplanması

    if i==1 or i==0:       # bir veya 0 rakamının özel faktöriyeli sabit 1.
        return 1
    
    else: 
        return i * factorialCalc(i-1)    # rekürsif

def permutationCalc(j,k):       #permütasyon iterasyon basamağı
    l = 0
    
    if k>j:        
        l = l
        
    else:      
        l = factorialCalc(j)/factorialCalc(j-k)   #faktöriyel kavramını çağırdığımız yer 
        return l

def checkGreaterNumbers(numOfClusterElmnt,digPer,foundOfNum):        # permütasyon koşullarını sağlayan rakamlar mı ? 
 
  if numOfClusterElmnt<digPer:
     print("Lütfen küme eleman sayısından fazla değer girmeyiniz!")
     digPer = int(input("Dizinin kaçlı permütasyonu hesaplanacak ?: "))     
     return  False
  elif numOfClusterElmnt-foundOfNum <digPer:
     return False
  else: 
    return True

print("Permütasyon hesabı için lütfen istenenleri giriniz.")   
numOfClusterElmnt = int(len(input("Permütasyon hesaplanacak kümenin elemanlarını (rakamları-karakterlerini) giriniz: ")))  #  ilk olarak dizinin uzunluğunu buluyoruz
digPer = int(input("Dizinin kaçlı permütasyonu hesaplanacak ?: "))             # Sonrasında kaçlı permütasyonu bulunması isteniyor
foundOfNum = int(len(input("Hangi iki rakamdan birisi veya ikisi bulunmalı?: ")))   # Hangi rakam(harf) veya rakamların(harf dizesi) sıralamada bulunması isteniyor   
while  checkGreaterNumbers(numOfClusterElmnt,digPer,foundOfNum) == False:
    checkGreaterNumbers(numOfClusterElmnt,digPer,foundOfNum)

print("Sonuç:", int(permutationCalc(numOfClusterElmnt,digPer) - permutationCalc(numOfClusterElmnt-foundOfNum,digPer) ))



Permütasyon hesabı için lütfen istenenleri giriniz.
ERROR! Session/line number was not unique in database. History logging moved to new session 103
Permütasyon hesaplanacak kümenin elemanlarını (rakamları-karakterlerini) giriniz: 123456
Dizinin kaçlı permütasyonu hesaplanacak ?: 4
Hangi iki rakamdan birisi veya ikisi bulunmalı?: 56
Sonuç: 336


# PERMÜTASYON
Permütasyon ,n pozitif tamsayı, r doğal sayı ve r ≤n olmak üzere, n elemanlı bir kümenin r elemanlı sıralı r’lilerine o kümenin r’li permütasyonu olarak tanımlayabiliriz.n tane farklı nesnemiz var. bu n tane nesneden r tanesini aynı anda seçiyor ve sıralıyoruz, yani sıralı r’li oluşturuyoruz, işte bu sıralamaya n’in r’li permütasyonu denir. Permütasyonda diziliş önemlidir.

P(n,r) = n!/(n−r)!  n adet nesneden r tanesini seçip  sıralamanın formülüdür.
Konuyu bir örnekle pekiştirelim.

Problem(Girdi) :Oğuz={1,2,3,4,5,6} kümesinin dörtlü permütasyonlarının kaçında 5 veya 6 bulunur ?

Çözüm : Öncelikle kümenin toplam eleman sayısının 4'lü permütasyonu bulunur.
P(6,4) = 360 

Akabinde 6 elemana sahip Oğuz kümesinin dört elemanlı alt kümelerinden 5 ve 6 rakamının bulunmaması durumu ayrıştırılır.
P(4,4) = 24
P(6,4) - P(4,4) = 360 - 24

Sonuç(Çıktı) : 360-24=336

